common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,961 @@
+    final boolean isTooLate = partitionInfoBeforeCheck == null;
+    // Check if the given block is already merged by checking the bitmap against the given map index
+    final AppShufflePartitionInfo partitionInfo = partitionInfoBeforeCheck != null
+      && partitionInfoBeforeCheck.mapTracker.contains(mapIndex) ? null : partitionInfoBeforeCheck;
+    if (partitionInfo != null) {
+      return new StreamCallbackImpl(this, msg, appShuffleId, reduceId, mapIndex, partitionInfo);
+    } else {
+      // For a duplicate block or a block which is late, respond back with a callback that handles
+      // them differently.
+      return new StreamCallbackWithID() {
+        @Override
+        public String getID() {
+          return msg.blockId;
+        }
+
+        @Override
+        public void onData(String streamId, ByteBuffer buf) {
+          // Ignore the requests. It reaches here either when a request is received after the
+          // shuffle file is finalized or when a request is for a duplicate block.
+        }
+
+        @Override
+        public void onComplete(String streamId) {
+          if (isTooLate) {
+            // Throw an exception here so the block data is drained from channel and server
+            // responds RpcFailure to the client.
+            throw new RuntimeException(String.format("Block %s %s", msg.blockId,
+              ErrorHandler.BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX));
+          }
+          // For duplicate block that is received before the shuffle merge finalizes, the
+          // server should respond success to the client.
+        }
+
+        @Override
+        public void onFailure(String streamId, Throwable cause) {
+        }
+      };
+    }
+  }
+
+  @SuppressWarnings("SynchronizationOnLocalVariableOrMethodParameter")
+  @Override
+  public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) throws IOException {
+    logger.info("Finalizing shuffle {} from Application {}.", msg.shuffleId, msg.appId);
+    Map<Integer, AppShufflePartitionInfo> shufflePartitions =
+      partitions.remove(new AppShuffleId(msg.appId, msg.shuffleId));
+    if (shufflePartitions == null || shufflePartitions.isEmpty()) {
+      logger.info("No partitions to finalize for shuffle {} from Application {}.",
+        msg.shuffleId, msg.appId);
+      return new MergeStatuses(msg.shuffleId, new RoaringBitmap[0], new int[0], new long[0]);
+    }
+    Collection<AppShufflePartitionInfo> partitions = shufflePartitions.values();
+    int totalPartitions = partitions.size();
+    RoaringBitmap[] bitmaps = new RoaringBitmap[totalPartitions];
+    int[] reduceIds = new int[totalPartitions];
+    long[] sizes = new long[totalPartitions];
+    Iterator<AppShufflePartitionInfo> partitionsIter = partitions.iterator();
+    int idx = 0;
+    while (partitionsIter.hasNext()) {
+      AppShufflePartitionInfo partition = partitionsIter.next();
+      partitionsIter.remove();
```

Comment:
Deletion should be inside the lock.
We only delete when we can finalize, and we can only finalize when we acquire the lock.

Reply:
`paritionsIter.remove()` should be after the data file is written so I have moved it after the files are closed. If it's before then there is a possibility that it will be removed from the shufflePartitions and data file doesn't exist. 
Another thread may reach line 132 where it will try creating a new `AppShufflePartitionInfo`